lambda.tf

```diff
@@ -1,130 +1,50 @@
 resource "aws_lambda_function" "lambda" {
-  count = "${! var.attach_vpc_config && ! var.attach_dead_letter_config ? 1 : 0}"
 
-  # ----------------------------------------------------------------------------
-  # IMPORTANT:
-  # Changes made to this resource should also be made to "lambda_with_*" below.
-  # ----------------------------------------------------------------------------
-
-  function_name                  = "${var.function_name}"
-  description                    = "${var.description}"
-  role                           = "${aws_iam_role.lambda.arn}"
-  handler                        = "${var.handler}"
-  memory_size                    = "${var.memory_size}"
-  reserved_concurrent_executions = "${var.reserved_concurrent_executions}"
-  runtime                        = "${var.runtime}"
-  layers                         = "${var.layers}"
-  timeout                        = "${local.timeout}"
-  publish                        = "${local.publish}"
-  tags                           = "${var.tags}"
+  function_name                  = var.function_name
+  description                    = var.description
+  role                           = aws_iam_role.lambda.arn
+  handler                        = var.handler
+  memory_size                    = var.memory_size
+  reserved_concurrent_executions = var.reserved_concurrent_executions
+  runtime                        = var.runtime
+  layers                         = var.layers
+  timeout                        = local.timeout
+  publish                        = local.publish
+  tags                           = var.tags
 
   # Use a generated filename to determine when the source code has changed.
 
-  filename   = "${lookup(data.external.built.result, "filename")}"
-  depends_on = ["null_resource.archive"]
-
-  # The aws_lambda_function resource has a schema for the environment
-  # variable, where the only acceptable values are:
-  #   a. Undefined
-  #   b. An empty list
-  #   c. A list containing 1 element: a map with a specific schema
-  # Use slice to get option "b" or "c" depending on whether a non-empty
-  # value was passed into this module.
-
-  environment = ["${slice( list(var.environment), 0, length(var.environment) == 0 ? 0 : 1 )}"]
-}
+  filename   = data.external.built.result.filename
+  depends_on = [null_resource.archive]
 
-# The vpc_config and dead_letter_config variables are lists of maps which,
-# due to a bug or missing feature of Terraform, do not work with computed
-# values. So here is a copy and paste of of the above resource for every
-# combination of these variables.
+  # Add dynamic blocks based on variables.
 
-resource "aws_lambda_function" "lambda_with_dl" {
-  count = "${var.attach_dead_letter_config && ! var.attach_vpc_config ? 1 : 0}"
-
-  dead_letter_config {
-    target_arn = "${var.dead_letter_config["target_arn"]}"
+  dynamic "dead_letter_config" {
```

Comment:
Woah. This is sweet.

Reply:
Gonna be a lot of folks needing to move resources in their state though, or let terraform resource cycle their lambdas. Still worth it I'd say.

Reply:
Yeah the dynamic blocks are good, although I still had to resort to a hack.

```terraform
  # this wouldn't work
  dynamic "dead_letter_config" {
    for_each = var.dead_letter_config == null ? [] : [var.dead_letter_config]
    content {
      target_arn = dead_letter_config.target_arn
    }
  }

  # so i used a dummy value of 1 and accessed var.dead_letter_config from inside the dynamic block
  dynamic "dead_letter_config" {
    for_each = var.dead_letter_config == null ? [] : [1]
    content {
      target_arn = var.dead_letter_config.target_arn
    }
  }
```

And yeah this will change the resource location for anyone using dead letter queues or VPC, but I think it's worth it too. I started to implement S3 as a source and that was going to add another 4 permutations, I think. Definitely nicer having just the one resource.

Reply:
I figured out what I was doing wrong after reading the docs again.

```terraform
  dynamic "dead_letter_config" {
    for_each = var.dead_letter_config == null ? [] : [var.dead_letter_config]
    content {
      target_arn = dead_letter_config.value.target_arn # <---- need to access .value to get the thing in from the list
    }
  }
```